File: 2025-06-June-GFG-POTD/June 29 - Split Array Largest Sum/Solution.java
class Solution {
     public int splitArray(int[] arr, int k) {
        int low = 0, high = 0;
        for (int num : arr) {
            low = Math.max(low, num); // the minimum possible answer
            high += num; // the maximum possible answer
        }

        while (low < high) {
            int mid = low + (high - low) / 2;
            if (canSplit(arr, k, mid)) {
                high = mid; // try for smaller max sum
            } else {
                low = mid + 1; // need larger max sum
            }
        }

        return low;
    }

    // Helper function to check if we can split the array with max sum <= maxSum
    private boolean canSplit(int[] arr, int k, int maxSum) {
        int currentSum = 0;
        int splits = 1;

        for (int num : arr) {
            if (currentSum + num > maxSum) {
                // Need to split here
                splits++;
                currentSum = num;
                if (splits > k) return false;
            } else {
                currentSum += num;
            }
        }

        return true;
    }
};

File: 2025-06-June-GFG-POTD/June 29 - Split Array Largest Sum/Solution.py
class Solution:
    def splitArray(self, arr, k):
        def can_split(arr, k, max_sum):
            current_sum = 0
            splits = 1
            for num in arr:
                if current_sum + num > max_sum:
                    splits += 1
                    current_sum = num
                    if splits > k:
                        return False
                else:
                    current_sum += num
            return True

        low = max(arr)
        high = sum(arr)

        while low < high:
            mid = (low + high) // 2
            if can_split(arr, k, mid):
                high = mid
            else:
                low = mid + 1

        return low
